test/docfx.Test/lib/GitHubUserCacheTest.cs

```diff
@@ -0,0 +1,201 @@
+                async (cache) => await cache.GetByCommit("alice_new@contoso.com", "owner", "name", "1"),
+                "[{'id':1,'login':'alice','name':'Alice','emails':['alice@contoso.com']}]".Replace("\'", "\""),
+                "[{'id':1,'login':'alice','name':'Alice','emails':['alice@contoso.com','alice_new@contoso.com']}]".Replace("\'", "\""),
+                0,
+                1);
+        }
+
+        [Fact]
+        public async Task GetUserByInvalidLoginAsync()
+        {
+            await VerifyCore(
+                async (cache) => await cache.GetByLogin("invalid"),
+                "[]",
+                "[{'login':'invalid','emails':[]}]".Replace("\'", "\""),
+                1,
+                0);
+        }
+
+        [Fact]
+        public async Task GetUserByInvalidLoginFromCacheAsync()
+        {
+            await VerifyCore(
+                async (cache) => await cache.GetByLogin("invalid"),
+                "[{'login':'invalid','emails':[]}]".Replace("\'", "\""),
+                "[{'login':'invalid','emails':[]}]".Replace("\'", "\""),
+                0,
+                0);
+        }
+
+        private async Task VerifyCore(Func<GitHubUserCache, Task> test, string cacheUsersJson, string expectedOutputCacheUsersJson, int expectedGetUserByLoginCall, int expectedGetLoginByCommitCall)
+        {
+            // Arrange
+            var (_, users) = JsonUtility.Deserialize<GitHubUser[]>(cacheUsersJson);
+            var cache = new GitHubUserCache(users, "cache.json", 7 * 24);
+            var accessor = new MockGitHubAccessor();
+            cache._getUserByLoginFromGitHub = accessor.GetUserByLogin;
+            cache._getLoginByCommitFromGitHub = accessor.GetLoginByCommit;
+
+            // Act
+            await test(cache);
+
+            // Assert
+            Assert.Equal(expectedGetUserByLoginCall, accessor.GetUserByLoginCallCount);
+            Assert.Equal(expectedGetLoginByCommitCall, accessor.GetLoginByCommitCallCount);
+            var (_, expectedUsers) = JsonUtility.Deserialize<GitHubUser[]>(expectedOutputCacheUsersJson);
+            AssertUsersEqual(expectedUsers, cache.Users.ToArray());
+        }
+
+        private void AssertUsersEqual(GitHubUser[] expectedUsers, GitHubUser[] actualUsers)
+        {
+            Assert.Equal(expectedUsers.Length, actualUsers.Length);
+            for (var i = 0; i < expectedUsers.Length; i++)
+            {
+                AssertUserEqual(expectedUsers[i], actualUsers[i]);
+            }
+        }
+
+        private void AssertUserEqual(GitHubUser expectedUser, GitHubUser actualUser)
+        {
+            Assert.Equal(expectedUser.Id, actualUser.Id);
+            Assert.Equal(expectedUser.Login, actualUser.Login);
+            Assert.Equal(expectedUser.Name, actualUser.Name);
+            Assert.True(expectedUser.Emails.OrderBy(u => u).SequenceEqual(actualUser.Emails.OrderBy(u => u)));
+        }
+
+        private sealed class MockGitHubAccessor
```

Comment:
have we considered using [moq](https://github.com/moq/moq)?

Reply:
Moq will introduce some test purpose interface to core logic, which we try to avoid.

---
In reply to: [226558147](https://github.com/dotnet/docfx/pull/3467#discussion_r226558147) [](ancestors = 226558147)